### PR TITLE
Feature dli new

### LINKED ISSUE #1256: Upgrade DLI queues to  the  go SDK of huaweicloud services
### Current Terraform Version

```
...
```

### Use-cases

Upgrade DLI queues to  the  go SDK of huaweicloud services

### PR DESCRIPTION
**Pull request by hmbig2**
**What this PR does / why we need it**:
Upgrade DLI queues to the go SDK of huaweicloud services
**Which issue this PR fixes**:
fixes #1256 

**Special notes for your reviewer**:

**Release note**:

```release-note

```

## PR Checklist

* [X] Tests added/passed.
* [X] Documentation updated.
* [X] Schema updated.

## Acceptance Steps Performed

```
make testacc TEST='./huaweicloud' TESTARGS='-run=TestAccDliQueueV1_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud -v -run=TestAccDliQueueV1_basic -timeout 360m -parallel 4
=== RUN   TestAccDliQueueV1_basic
=== PAUSE TestAccDliQueueV1_basic
=== CONT  TestAccDliQueueV1_basic
--- PASS: TestAccDliQueueV1_basic (18.61s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud       19.318s
```

### CODE REVIEW

File: huaweicloud/services/dli/resource_huaweicloud_dli_queue_v1.go [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+var regexp4Name = regexp.MustCompile(`^[a-z0-9_]{1,128}$`)
+
+func ResourceDliQueueV1() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceDliQueueCreate,
+		Read:   resourceDliQueueV1Read,
+		Delete: resourceDliQueueV1Delete,
+
+		Schema: map[string]*schema.Schema{
+			"name": {
```
Comment (chengxiangdong) [inline huaweicloud/services/dli/resource_huaweicloud_dli_queue_v1.go]: `region` is required and is a reserved field.

File: docs/resources/dli_queue.md [hunk 2]
```diff
@@ -14,31 +16,56 @@ This is an alternative to `huaweicloud_dli_queue`
 ```hcl
 resource "huaweicloud_dli_queue" "queue" {
   name     = "terraform_dli_queue_test"
-  cu_count = 4
+  cu_count = 16
+  tags     = {
+    k1   =   "1"
+  }
 }
```
Comment (chengxiangdong) [inline docs/resources/dli_queue.md]: wrong format

File: docs/resources/dli_queue.md [hunk 3]
```diff
@@ -14,31 +16,56 @@ This is an alternative to `huaweicloud_dli_queue`
...
-
-* `vpc_cidr` - (Optional, String, ForceNew) VPC CIDR. Changing this parameter will create a new resource.
-
+* `name` - (Required, String, ForceNew) Name of a queue. Name of a newly created resource queue. 
+    The name can contain only digits, letters, and underscores (_), 
+    but cannot contain only digits or start with an underscore (_).
+    Length range: 1 to 128 characters. Changing this parameter will create a new resource.
+
+* `queue_type` - (Optional, String, ForceNew) Indicates the queue type. 
+    Changing this parameter will create a new resource. The options are as follows:
```
Comment (chengxiangdong) [inline docs/resources/dli_queue.md:39]: Suggest to write like "Indicates the queue type. Defaults to `sql`.Valid values are:`sql` and `general`." 

The style is consistent with other docs, more concise.

File: docs/resources/dli_queue.md [hunk 4]
```diff
@@ -13,32 +15,58 @@ This is an alternative to `huaweicloud_dli_queue`
 
 ```hcl
 resource "huaweicloud_dli_queue" "queue" {
-  name     = "terraform_dli_queue_test"
-  cu_count = 4
+  name          = "terraform_dli_queue_test"
+  cu_count      = 16
+  resource_mode = 0
+  tags          = {
+    "k1" = "1"
```
Comment (chengxiangdong) [inline docs/resources/dli_queue.md]: `"k1" = "1"` -> `k1 = "1"`

File: huaweicloud/services/acceptance/dli/resource_huaweicloud_dli_queue_v1_test.go [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+					testAccCheckDliQueueV1Exists(resourceName),
+				),
+			},
+		},
+	})
+}
+
+func testAccDliQueueV1_basic(rName string) string {
+	return fmt.Sprintf(`
+resource "huaweicloud_dli_queue_v1" "test" {
```
Comment (chengxiangdong) [inline huaweicloud/services/acceptance/dli/resource_huaweicloud_dli_queue_v1_test.go:40]: wrong format

File: huaweicloud/services/acceptance/dli/resource_huaweicloud_dli_queue_v1_test.go [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+	name             =  "%s"
+	cu_count         =   16
+	resource_mode    =   0
+	tags             = {
+		k1    =    "1"
+	}
+}`, rName)
```
Comment (chengxiangdong) [inline huaweicloud/services/acceptance/dli/resource_huaweicloud_dli_queue_v1_test.go]: wrong format

File: huaweicloud/services/acceptance/dli/resource_huaweicloud_dli_queue_v1_test.go [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+func testAccCheckDliQueueV1Destroy(s *terraform.State) error {
+	config := act.TestAccProvider.Meta().(*config.Config)
+	client, err := config.DliV1Client(act.HW_REGION_NAME)
+	if err != nil {
+		return fmtp.Errorf("Error creating sdk client, err=%s", err)
```
Comment (chengxiangdong) [inline huaweicloud/services/acceptance/dli/resource_huaweicloud_dli_queue_v1_test.go]: Suggest to print which sdk client error.

File: huaweicloud/services/dli/resource_huaweicloud_dli_queue_v1.go [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+			},
+		},
+	}
+}
+
+func resourceDliQueueCreate(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*config.Config)
+	dliClient, err := config.DliV1Client(config.GetRegion(d))
+	if err != nil {
+		return fmt.Errorf("creating dli client failed: %s", err)
```
Comment (chengxiangdong) [inline huaweicloud/services/dli/resource_huaweicloud_dli_queue_v1.go:115]: `fmt` -> `fmtp`

Comment (chengxiangdong) [inline huaweicloud/services/dli/resource_huaweicloud_dli_queue_v1.go:115]: `fmt` -> `fmtp`

File: huaweicloud/services/dli/resource_huaweicloud_dli_queue_v1.go [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+		Platform:            d.Get("platform").(string),
+		ResourceMode:        d.Get("resource_mode").(int),
+		Feature:             d.Get("feature").(string),
+		Tags:                assembleTagsFromRecource("tags", d),
+	}
+
+	logp.Printf("[DEBUG] create dli queues using paramaters: %+v", createOpts)
+	createResult := queues.Create(dliClient, createOpts)
+	if createResult.Err != nil {
+		return fmt.Errorf("create dli queues failed: %s", createResult.Err)
```
Comment (chengxiangdong) [inline huaweicloud/services/dli/resource_huaweicloud_dli_queue_v1.go:136]: `fmt` -> `fmtp`

File: huaweicloud/services/dli/resource_huaweicloud_dli_queue_v1.go (hunk 8 above)
Comment (chengxiangdong) [inline huaweicloud/services/dli/resource_huaweicloud_dli_queue_v1.go:115]: `fmt` -> `fmtp`

Comment (chengxiangdong) [inline huaweicloud/services/dli/resource_huaweicloud_dli_queue_v1.go]: `fmt` -> `fmtp` pls, total 4 places

File: huaweicloud/services/dli/resource_huaweicloud_dli_queue_v1.go [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+		return fmtp.Errorf("error creating DliV1Client, err=%s", err)
+	}
+
+	queueName := d.Get("name").(string)
+
+	queryOpts := queues.ListOpts{
+		QueueType: d.Get("queue_type").(string),
+	}
+
+	log.Printf("[DEBUG] create dli queues using paramaters: %+v", queryOpts)
```
Comment (chengxiangdong) [inline huaweicloud/services/dli/resource_huaweicloud_dli_queue_v1.go]: `log` -> `logp` pls, total 3 places

File: huaweicloud/services/dli/resource_huaweicloud_dli_queue_v1.go [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+
+	logp.Printf("[DEBUG] create dli queues queueName: %s", queueName)
+	createOpts := queues.CreateOpts{
+		QueueName:           queueName,
+		QueueType:           d.Get("queue_type").(string),
+		Description:         d.Get("description").(string),
+		CuCount:             d.Get("cu_count").(int),
+		EnterpriseProjectId: config.GetEnterpriseProjectID(d),
+		Platform:            d.Get("platform").(string),
+		ResourceMode:        d.Get("resource_mode").(int),
```
Comment (chengxiangdong) [inline huaweicloud/services/dli/resource_huaweicloud_dli_queue_v1.go]: The type of `ResourceMode` is `int`, when the value is `0` this parameter will be discarded during conversion process of sending the http request to the server.

Maybe the service will set it to `0`, but it is still recommended to change the type to `*int`.

File: docs/resources/dli_queue.md (hunk 2 above)
Comment (hmbig2) [reply docs/resources/dli_queue.md]: fixed

File: huaweicloud/services/dli/resource_huaweicloud_dli_queue_v1.go (hunk 11 above)
Comment (hmbig2) [reply huaweicloud/services/dli/resource_huaweicloud_dli_queue_v1.go]: hava change the json tag in sdk. Always serialize this field to zero when it's nil

**Review by chengxiangdong — COMMENTED** [review]
all commits has resolved
CC @ShiChangkuo

File: docs/resources/dli_queue.md [hunk 12]
```diff
@@ -13,32 +15,63 @@ This is an alternative to `huaweicloud_dli_queue`
...
+    - x86_64 : default value
+    - aarch64
+
+* `resource_mode` - (Optional, String, ForceNew) Queue resource mode. 
+  Changing this parameter will create a new resource. 
+  The options are as follows: 
+  - 0: indicates the shared resource mode.
+  - 1: indicates the exclusive resource mode. 
+
+* `feature` - (Optional, String, ForceNew)Indicates the queue feature.
```
Comment (ShiChangkuo) [inline docs/resources/dli_queue.md:65]: I can not find the option in console, shall we remove it?

File: docs/resources/dli_queue.md [hunk 13]
```diff
@@ -2,9 +2,11 @@
 subcategory: "Data Lake Insight (DLI)"
 ---
 
-# huaweicloud\_dli\_queue
+# huaweicloud_dli_queue
 
 DLI Queue management
+Allows you to create a queue. The queue will be bound to specified compute resources.
+
 This is an alternative to `huaweicloud_dli_queue`
```
Comment (ShiChangkuo) [inline docs/resources/dli_queue.md]: please follow the format: Manages a xxxx resource within HuaweiCloud.

File: huaweicloud/services/dli/resource_huaweicloud_dli_queue_v1.go [hunk 14]
```diff
@@ -0,0 +1,235 @@
+package dli
+
+import (
+	"fmt"
```
Comment (ShiChangkuo) [inline huaweicloud/services/dli/resource_huaweicloud_dli_queue_v1.go]: should used **fmtp**

File: huaweicloud/services/dli/resource_huaweicloud_dli_queue_v1.go [hunk 15]
```diff
@@ -0,0 +1,235 @@
...
+				Computed: true,
+				Optional: true,
+				ForceNew: true,
+			},
+
+			"cu_count": {
+				Type:         schema.TypeInt,
+				Required:     true,
+				ForceNew:     true,
+				ValidateFunc: validation.IntInSlice([]int{CU_16, CU_64, CU_256}),
```
Comment (ShiChangkuo) [inline huaweicloud/services/dli/resource_huaweicloud_dli_queue_v1.go]: please add 512 into the slice.

File: huaweicloud/services/dli/resource_huaweicloud_dli_queue_v1.go [hunk 16]
```diff
@@ -0,0 +1,235 @@
...
+
+const CU_16, CU_64, CU_256 = 16, 64, 256
+const RESOURCE_MODE_SHARED, RESOURCE_MODE_EXCLUSIVE = 0, 1
+
+func ResourceDliQueueV1() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceDliQueueCreate,
+		Read:   resourceDliQueueV1Read,
+		Delete: resourceDliQueueV1Delete,
+
```
Comment (ShiChangkuo) [inline huaweicloud/services/dli/resource_huaweicloud_dli_queue_v1.go]: suggest to support the **import** function.

File: huaweicloud/services/dli/resource_huaweicloud_dli_queue_v1.go [hunk 17]
```diff
@@ -0,0 +1,235 @@
...
+	config := meta.(*config.Config)
+	client, err := config.DliV1Client(config.GetRegion(d))
+	if err != nil {
+		return fmtp.Errorf("error creating DliV1Client, err=%s", err)
+	}
+
+	queueName := d.Get("name").(string)
+	logp.Printf("[DEBUG] Deleting dli Queue %q", d.Id())
+
+	result := queues.Delete(client, queueName)
```
Comment (ShiChangkuo) [inline huaweicloud/services/dli/resource_huaweicloud_dli_queue_v1.go]: seems we can use d.Id() to replace `queueName`

File: docs/resources/dli_queue.md [hunk 18]
```diff
@@ -13,32 +15,63 @@ This is an alternative to `huaweicloud_dli_queue`
 
 ```hcl
 resource "huaweicloud_dli_queue" "queue" {
-  name     = "terraform_dli_queue_test"
-  cu_count = 4
+  name          = "terraform_dli_queue_test"
+  cu_count      = 16
+  resource_mode = 0
```
Comment (ShiChangkuo) [inline docs/resources/dli_queue.md]: there is unnecessary to specifie resource_mode, as it is optional and defaults to 0.

File: docs/resources/dli_queue.md [hunk 19]
```diff
@@ -13,32 +15,63 @@ This is an alternative to `huaweicloud_dli_queue`
...
+  
+  tags = {
+    k1 = "1"
```
Comment (ShiChangkuo) [inline docs/resources/dli_queue.md]: please use more meaningful tags as the example, like
```
  tags = {
    foo = "bar"
    key = "value"
  }
```

File: huaweicloud/services/dli/resource_huaweicloud_dli_queue_v1.go [hunk 20]
```diff
@@ -0,0 +1,235 @@
...
+const CU_16, CU_64, CU_256 = 16, 64, 256
+const RESOURCE_MODE_SHARED, RESOURCE_MODE_EXCLUSIVE = 0, 1
+
+func ResourceDliQueueV1() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceDliQueueCreate,
+		Read:   resourceDliQueueV1Read,
+		Delete: resourceDliQueueV1Delete,
+
+		Schema: map[string]*schema.Schema{
```
Comment (ShiChangkuo) [inline huaweicloud/services/dli/resource_huaweicloud_dli_queue_v1.go]: We should keep `management_subnet cidr`, `subnet cidr` and `vpc cidr` compatible with previous versions.

**Review by ShiChangkuo — CHANGES_REQUESTED** [review]
@hmbig2 please remove the files in *huaweicloud* dir if you want to make a new pakage for DLI.

BTW, it's not recommended to use file names and function names with version numbers（_v1）

File: huaweicloud/services/dli/resource_huaweicloud_dli_queue_v1.go (hunk 15 above)
Comment (hmbig2) [reply huaweicloud/services/dli/resource_huaweicloud_dli_queue_v1.go]: 516 is not avaiable option in api docs

File: huaweicloud/services/dli/resource_huaweicloud_dli_queue_v1.go (hunk 16 above)
Comment (hmbig2) [reply huaweicloud/services/dli/resource_huaweicloud_dli_queue_v1.go]: OK,  add import ,update  in next release version

File: huaweicloud/services/acceptance/dli/resource_huaweicloud_dli_queue_v1_test.go [hunk 21]
```diff
@@ -0,0 +1,99 @@
...
+
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:     func() { act.TestAccPreCheck(t) },
+		Providers:    act.TestAccProviders,
+		CheckDestroy: testAccCheckDliQueueV1Destroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccDliQueueV1_basic(rName),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckDliQueueV1Exists(resourceName),
```
Comment (chengxiangdong) [inline huaweicloud/services/acceptance/dli/resource_huaweicloud_dli_queue_v1_test.go]: Unit test verification is not sufficient, suggest to add some attributes verification.

**Review by ShiChangkuo — APPROVED** [review]

**Pull request merged**